server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java

```diff
@@ -298,8 +298,12 @@ private void tryGetFromTranslog(GetRequest request, IndexShard indexShard, Clust
     }
 
     static DiscoveryNode getCurrentNodeOfPrimary(ClusterState clusterState, ShardId shardId) {
-        var shardRoutingTable = clusterState.routingTable().shardRoutingTable(shardId);
-        if (shardRoutingTable.primaryShard() == null || shardRoutingTable.primaryShard().active() == false) {
+        var indexRoutingTable = clusterState.routingTable().index(shardId.getIndex());
+        if (indexRoutingTable == null) {
+            return null;
```

Comment:
I think need a different exception here? I would expect `IndexNotFoundException`, not `NoShardAvailableActionException`. The former is a 4xx error, the latter a 5xx error.